pkg/scheduler/preemption/preemption.go

```diff
@@ -572,58 +562,21 @@ func workloadUsesResources(wl *workload.Info, resPerFlv resourcesPerFlavor) bool
 // requestable resources and simulated usage of the ClusterQueue and its cohort,
 // if it belongs to one.
 func workloadFits(wlReq resources.FlavorResourceQuantitiesFlat, cq *cache.ClusterQueueSnapshot, allowBorrowing bool) bool {
-	for _, rg := range cq.ResourceGroups {
-		for _, fName := range rg.Flavors {
-			cqResUsage := cq.Usage[fName]
-			for rName := range rg.CoveredResources {
-				resource := cq.QuotaFor(resources.FlavorResource{Flavor: fName, Resource: rName})
-				rReq, found := wlReq[resources.FlavorResource{Flavor: fName, Resource: rName}]
-				if !found {
-					// Workload doesn't request this FlavorResource.
-					continue
-				}
-
-				if cq.Cohort == nil || !allowBorrowing {
-					if cqResUsage[rName]+rReq > resource.Nominal {
-						return false
-					}
-				} else {
-					// When resource.BorrowingLimit == nil there is no borrowing
-					// limit, so we can skip the check.
-					if resource.BorrowingLimit != nil {
-						if cqResUsage[rName]+rReq > resource.Nominal+*resource.BorrowingLimit {
-							return false
-						}
-					}
-				}
-
-				if cq.Cohort != nil {
-					cohortResUsage := cq.UsedCohortQuota(fName, rName)
-					requestableQuota := cq.RequestableCohortQuota(fName, rName)
-					if cohortResUsage+rReq > requestableQuota {
-						return false
-					}
-				}
-			}
+	for fr, v := range wlReq {
+		if cq.BorrowingWith(fr, v) && !allowBorrowing {
+			return false
+		}
+		if v > cq.Available(fr) {
+			return false
 		}
 	}
 	return true
 }
 
 func queueUnderNominalInResourcesNeedingPreemption(resPerFlv resourcesPerFlavor, cq *cache.ClusterQueueSnapshot) bool {
-	for _, rg := range cq.ResourceGroups {
-		for _, fName := range rg.Flavors {
-			flvReq, found := resPerFlv[fName]
-			if !found {
-				// Workload doesn't request this flavor.
-				continue
-			}
-			cqResUsage := cq.Usage[fName]
-			for rName := range flvReq {
-				if cqResUsage[rName] >= cq.QuotaFor(resources.FlavorResource{Flavor: fName, Resource: rName}).Nominal {
```

Comment:
note to reviewers: this `>=` turns into a `>`. I think this is correct (as we want to make sure we're within nominal quota), but please scrutinize it @alculquicondor

Reply:
Actually, a more accurate check would be whether the usage plus the resources for the incoming workload would be borrowing.

If it's not borrowing, then this means that this CQ is preempting to reclaim quota, then it is allowed to preempt other workloads in the cohort. Otherwise, it should only be allowed to preempt workloads within its CQ or those below the threshold.

@mimowo wdyt? This was added in #1979 and reused in #2110

Reply:
If there's a bug in the accounting, let's fix it in a follow-up PR, as this PR is intended to cleanup existing code without any change in behavior

Reply:
> note to reviewers: this >= turns into a >. I think this is correct (as we want to make sure we're within nominal quota), but please scrutinize it @alculquicondor

I think `!queueUnderNominalInResourcesNeedingPreemption <> cqIsBorrowing`. For example if you have one resource, and `cqResUsage[rName] == Nominal` then the answer is "false", so `!queueUnderNominalInResourcesNeedingPreemption is true`, while `cqIsBorrowing` is false.

> @mimowo wdyt? This was added in https://github.com/kubernetes-sigs/kueue/pull/1979 and reused in https://github.com/kubernetes-sigs/kueue/pull/2110

ack,  I remember this was quite complex, so I would prefer to keep the logic in this PR, and do a dedicated one for fix if needed.  I will yet look a bit into this.

Reply:
Folded the two functions into one https://github.com/kubernetes-sigs/kueue/pull/2595#discussion_r1680826735

Reply:
I've met with @gabesaba and agreed that he will try to adjust this refactoring PR not to change any logic.We will have a dedicated follow up to adjust the logic if needed. We will then consider changing <, with <=, or the idea from https://github.com/kubernetes-sigs/kueue/pull/2595#discussion_r1676386670.

Reply:
Done. Unfolded the functions, and added back in the cohort == nil check.